File: app/src/main/java/io/github/wulkanowy/ui/widgets/TimetableWidgetFactory.java
package io.github.wulkanowy.ui.widgets;

import android.content.Context;
import android.content.Intent;
import android.graphics.Paint;
import android.view.View;
import android.widget.AdapterView;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;

import io.github.wulkanowy.R;
import io.github.wulkanowy.data.RepositoryContract;
import io.github.wulkanowy.data.db.dao.entities.TimetableLesson;
import io.github.wulkanowy.data.db.dao.entities.Week;
import io.github.wulkanowy.utils.TimeUtils;

public class TimetableWidgetFactory implements RemoteViewsService.RemoteViewsFactory {

    private final Context context;

    private List<TimetableLesson> lessonList = new ArrayList<>();

    private final RepositoryContract repository;

    public TimetableWidgetFactory(Context context, RepositoryContract repository) {
        this.context = context;
        this.repository = repository;
    }


    @Override
    public void onCreate() {
        // do nothing
    }

    @Override
    public void onDataSetChanged() {
        lessonList = new ArrayList<>();

        if (repository.getSharedRepo().isUserLoggedIn()) {

            Week week = repository.getDbRepo().getWeek(TimeUtils.getDateOfCurrentMonday(true));
            int valueOfDay = TimeUtils.getTodayOrNextDayValue(repository.getSharedRepo().getTimetableWidgetState());

            if (valueOfDay != 5 && valueOfDay != 6 && week != null) {
                week.resetDayList();
                lessonList = week.getDayList().get(valueOfDay).getTimetableLessons();
            }
        }
    }

    @Override
    public void onDestroy() {
        // do nothing
    }

    @Override
    public int getCount() {
        return lessonList.size();
    }

    @Override
    public RemoteViews getViewAt(int position) {
        if (position == AdapterView.INVALID_POSITION) {
            return null;
        }

        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.timetable_widget_item);
        views.setTextViewText(R.id.timetable_widget_item_subject, getSubjectName(position));
        views.setTextViewText(R.id.timetable_widget_item_time, getTimeText(position));
        views.setTextViewText(R.id.timetable_widget_item_room, getRoomText(position));

        if (!getDescriptionText(position).isEmpty()) {
            views.setViewVisibility(R.id.timetable_widget_item_description, View.VISIBLE);
            views.setTextViewText(R.id.timetable_widget_item_description, getDescriptionText(position));
        } else {
            views.setViewVisibility(R.id.timetable_widget_item_description, View.GONE);
        }

        if (lessonList.get(position).getMovedOrCanceled()) {
            views.setInt(R.id.timetable_widget_item_subject, "setPaintFlags",
                    Paint.STRIKE_THRU_TEXT_FLAG | Paint.ANTI_ALIAS_FLAG);
        } else {
            views.setInt(R.id.timetable_widget_item_subject, "setPaintFlags",
                    Paint.ANTI_ALIAS_FLAG);
        }

        views.setOnClickFillInIntent(R.id.timetable_widget_item_container, new Intent());
        return views;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    private String getSubjectName(int position) {
        return lessonList.get(position).getSubject();
    }

    private String getRoomText(int position) {
        TimetableLesson lesson = lessonList.get(position);
        if (!lesson.getRoom().isEmpty()) {
            return context.getString(R.string.timetable_dialog_room) + " " + lesson.getRoom();
        }
        return lesson.getRoom();
    }

    private String getTimeText(int position) {
        TimetableLesson lesson = lessonList.get(position);
        return lesson.getStartTime() + " - " + lesson.getEndTime();
    }

    private String getDescriptionText(int position) {
        return StringUtils.capitalize(lessonList.get(position).getDescription());
    }

}

File: app/src/main/java/io/github/wulkanowy/ui/widgets/timetable/TimetableWidgetFactory.kt
package io.github.wulkanowy.ui.widgets.timetable

import android.content.Context
import android.content.Intent
import android.graphics.Paint.ANTI_ALIAS_FLAG
import android.graphics.Paint.STRIKE_THRU_TEXT_FLAG
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.AdapterView.INVALID_POSITION
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import io.github.wulkanowy.R
import io.github.wulkanowy.data.db.SharedPrefHelper
import io.github.wulkanowy.data.db.entities.Timetable
import io.github.wulkanowy.data.repositories.SemesterRepository
import io.github.wulkanowy.data.repositories.StudentRepository
import io.github.wulkanowy.data.repositories.TimetableRepository
import io.github.wulkanowy.utils.toFormattedString
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import org.threeten.bp.LocalDate
import timber.log.Timber

class TimetableWidgetFactory(
    private val timetableRepository: TimetableRepository,
    private val studentRepository: StudentRepository,
    private val semesterRepository: SemesterRepository,
    private val sharedPref: SharedPrefHelper,
    private val context: Context,
    private val intent: Intent?
) : RemoteViewsService.RemoteViewsFactory {

    private val disposable: CompositeDisposable = CompositeDisposable()

    private var lessons = emptyList<Timetable>()

    override fun getLoadingView() = null

    override fun hasStableIds() = true

    override fun getCount() = lessons.size

    override fun getViewTypeCount() = 1

    override fun getItemId(position: Int) = position.toLong()

    override fun onCreate() {}

    override fun onDataSetChanged() {
        intent?.action?.let { LocalDate.ofEpochDay(sharedPref.getLong(it, 0)) }
            ?.let { date ->
                disposable.add(studentRepository.isStudentSaved()
                    .flatMap {
                        if (it) studentRepository.getCurrentStudent()
                        else Single.error(IllegalArgumentException("No saved students"))
                    }.flatMap { semesterRepository.getCurrentSemester(it) }
                    .flatMap { timetableRepository.getTimetable(it, date, date) }
                    .map { item -> item.sortedBy { it.number } }
                    .subscribe({ lessons = it })
                    { Timber.e(it, "An error has occurred while downloading data for the widget") })
            }
    }

    override fun getViewAt(position: Int): RemoteViews? {
        if (position == INVALID_POSITION) return null

        return RemoteViews(context.packageName, R.layout.item_widget_timetable).apply {
            lessons[position].let {
                setTextViewText(R.id.timetableWidgetItemSubject, it.subject)
                setTextViewText(R.id.timetableWidgetItemNumber, it.number.toString())
                setTextViewText(R.id.timetableWidgetItemTime, it.start.toFormattedString("HH:mm") +
                    " - ${it.end.toFormattedString("HH:mm")}")

                if (it.room.isNotBlank()) {
                    setTextViewText(R.id.timetableWidgetItemRoom, "${context.getString(R.string.timetable_room)} ${it.room}")
                } else setTextViewText(R.id.timetableWidgetItemRoom, "")

                if (it.info.isNotBlank()) {
                    setViewVisibility(R.id.timetableWidgetItemDescription, VISIBLE)
                    setTextViewText(R.id.timetableWidgetItemDescription, it.info.capitalize())
                } else setViewVisibility(R.id.timetableWidgetItemDescription, GONE)

                if (it.canceled) {
                    setInt(R.id.timetableWidgetItemSubject, "setPaintFlags",
                        STRIKE_THRU_TEXT_FLAG or ANTI_ALIAS_FLAG)
                } else {
                    setInt(R.id.timetableWidgetItemSubject, "setPaintFlags", ANTI_ALIAS_FLAG)
                }
            }
            setOnClickFillInIntent(R.id.timetableWidgetItemContainer, Intent())
        }
    }

    override fun onDestroy() {
        disposable.clear()
    }
}
